File: challenge-361/perl/ch-2.pl
#!/usr/bin/perl
use v5.38;
use warnings;

use List::Util qw(reduce all);

# Read the matrix from stdin, one row per line, values space-separated.
my @party = map { [ split ] } <STDIN>;

say find_celebrity(@party);

# Find the celebrity using a two-phase O(n) algorithm:
#   1. Narrow down to a single candidate by eliminating non-celebrities.
#   2. Verify the candidate is known by everyone and knows nobody.
sub find_celebrity {
    my @party = @_;
    my $n = scalar @party;

    # Phase 1: find the candidate.
    # If A knows B, A is not the celebrity. If A doesn't know B, B is not the celebrity.
    my $candidate = reduce { $party[$a][$b] ? $b : $a } 0 .. $n - 1;

    # Phase 2: verify the candidate.
    # The candidate must not know anyone, and everyone must know the candidate.
    my $valid = all { $_ == $candidate || (!$party[$candidate][$_] && $party[$_][$candidate]) } 0 .. $n - 1;

    return $valid ? $candidate : -1;
}

__END__

File: challenge-361/python/ch-2.py
import sys
from functools import reduce

# Read the matrix from stdin, one row per line, values space-separated.
party = [[int(x) for x in line.split()] for line in sys.stdin]

def find_celebrity(party):
    n = len(party)

    # Phase 1: find the candidate.
    # If A knows B, A is not the celebrity. If A doesn't know B, B is not the celebrity.
    candidate = reduce(lambda a, b: b if party[a][b] else a, range(n))

    # Phase 2: verify the candidate.
    # The candidate must not know anyone, and everyone must know the candidate.
    valid = all(i == candidate or (not party[candidate][i] and party[i][candidate]) for i in range(n))

    return candidate if valid else -1

print(find_celebrity(party))

